crates/astria-conductor/src/celestia/verify.rs

```diff
@@ -368,3 +385,149 @@ fn should_retry(error: &tendermint_rpc::Error) -> bool {
+            })
+            .await?
+        {
+            VerificationResponse::Commit(commit) => Ok(commit),
+            other => panic!("expected VerificationResponse::Commit, got {other:?}"),
+        }
+    }
+
+    async fn get_validators(
+        mut self,
+        prev_height: SequencerHeight,
+        height: SequencerHeight,
+    ) -> Result<Box<tendermint_rpc::endpoint::validators::Response>, BoxError> {
+        // allow: it is desired that the wildcard matches all future added variants because
+        // this call must only return a a single specific variant, panicking otherwise.
+        #[allow(clippy::match_wildcard_for_single_variants)]
+        match self
+            .inner
+            .ready()
+            .await?
+            .call(VerificationRequest::Validators {
+                prev_height,
+                height,
+            })
+            .await?
+        {
+            VerificationResponse::Validators(validators) => Ok(validators),
+            other => panic!("expected VerificationResponse::Validators, got {other:?}"),
+        }
+    }
+
+    fn new(client: SequencerClient) -> Self {
+        // XXX: the construction in here is a bit strange:
+        // the straight forward way to create a type-erased tower service is to use
+        // ServiceBuilder::boxed_clone().
+        //
+        // However, this gives a BoxCloneService which is always Clone + Send, but !Sync.
+        // Therefore we can't use the ServiceBuilder::buffer adapter.
+        //
+        // We can however work around it: ServiceBuilder::boxed gives a BoxService, which is
+        // Send + Sync, but not Clone. We then then manually evoke Buffer::new to create a
+        // a Buffer<BoxService>, which is Send + Sync + Clone.
+        let requests_per_minute = 100usize;
+        let service = tower::ServiceBuilder::new()
+            .boxed()
+            .rate_limit(
+                requests_per_minute.try_into().expect(
+                    "the requests per minute should be set to a reasonable size so that \
+                     conversion to u64 works on any platform",
+                ),
+                Duration::from_secs(60),
+            )
+            .service_fn(move |req: VerificationRequest| {
+                let client = client.clone();
+                async move {
+                    match req {
+                        VerificationRequest::Commit {
+                            height,
+                        } => fetch_commit_with_retry(client, height).await,
+                        VerificationRequest::Validators {
+                            prev_height,
+                            height,
+                        } => fetch_validators_with_retry(client, prev_height, height).await,
+                    }
+                }
+            });
+        // XXX: This number is arbitarily set to the same number os the rate-limit. Does that
+        // make sense? Should the number be set higher?
```

Comment:
I think it probably does make sense since it should be [at least the maximum number of concurrent requests the `Buffer` will see](https://docs.rs/tower/latest/tower/buffer/struct.Buffer.html#a-note-on-choosing-a-bound), although I also can't see an issue with setting it higher (within reason).